qiskit_dynamics/solvers/solver_functions.py

```diff
@@ -284,6 +310,28 @@ def rhs(t, y):
 
     if method == "scipy_expm":
         results = scipy_expm_solver(solver_generator, t_span, y0, t_eval=t_eval, **kwargs)
+    elif "lanczos_diag" in method:
```

Comment:
This block should be cleaned up/compartmentalized to keep this block of `if/else` statements in `solve_lmde` readable.
- First, I think the check using `is_hermitian` can be removed. I know we've discussed this before but I've generally tried to avoid evaluating the generator outside of solving (I don't believe it's done anywhere else). So long as the fact that this method is restricted to anti-Hermitian matrices is properly documented (it is) I think it's fine to just remove this.
- The other warnings/errors should be moved into a separate function, called something like `_lanczos_validation(generator)`, just to minimize the number of lines appearing in this block of code. This block can then be shortened to:
```
_lanczos_validation(generator)

if method == "lanczos_diag":
    results = ...
elif method == "jax_lanczos_diag":
    results = ...
```
- I also think the logic of the warnings/errors can be simplified as follows. Not a major change, but the error should be raised first if its going to be raised, and I prefer checking if its `not` a HamiltonianModel when raising the error, rather than checking if it's any of the others (in case more are added in the future).
```
if isinstance(generator, BaseGeneratorModel):
    if not isinstance(generator, HamiltonianModel):
          raise QiskitError(...)
    
    if "sparse" not in generator.evaluation_mode:
         warn(...)
```
Also I'd change the warning to say they *should* be used in sparse mode, rather than that they *must*.
- The last question I have related to the warning is: why `stacklevel=5` ? For similar warnings I've used `stacklevel=2` in the past.

Reply:
I was just trying out different stacklevels as i wasn't sure which was appropriate. No particular reason for `stacklevel=5` changed it to 2.